### DISCUSSION TITLE
config.nu location (macos)

### DISCUSSION THREAD

**Discussion by jesper-olsen** [nushell/nushell#14663] [Q&A] [answered]
Installed nushell v. 0.101 today (macos 15.1.1).

Seems nushell is using
~/Library/Application Support/nushell/config.nu  

when other applications (gcloud, iterm2, kitty, ...) generally use ~/.config

Should it not use the same by default?

**Comment by fdncred** [comment]
No, but there are ways to change it.
1. You can symlink to that folder
2. You can use XDG_CONFIG_HOME as our docs indicate

**Comment by ghost** [comment, marked as answer]
Each operating system has its own location where it recommends storing configuration files.  
Linux usually uses `~/.config` and macos recommends `~/Library/Application Support`.  
Many application developers (like gclould, iterm2, kitty, etc) just use the linux one due to either personal preference or because they didn't think about it when initially adding macos support.

For nushell this has been discussed quite a few times over the years and during [a poll last year](https://github.com/nushell/nushell/discussions/9019) the majority preferred `~/.config` over macos' location.  
However it seems to not been implemented (yet?).

**Comment by kaidev0711** [comment]
Startup variables in docs nushell XDG_CONFIG_HOME

**Comment by tmillr** [comment]
Take it from a Mac user, I agree with OP here and I do not like nushell's default on macOS. The canonical/conventional location for unix tools is `~/.config`, even on macOS (which ***is*** a POSIX-conforming, unix-like OS based on Darwin and BSD). The only real exceptions to this are GUI applications and macOS app bundles which use `~/Library/Application Support`, or (less often) very old tools (e.g. before 2010) whose inception predates the XDG spec. The former of which the config location is mostly irrelevant as configuring ***those*** type of apps is done through a GUI, and not by navigating to the file/filepath in a text editor.

This really should be changed, because sane/expected defaults go a very long way to making a piece of software user-friendly and easy on new users. Assuming the user doesn't modify the default (as they are more or less forced to), by using this awkward/unexpected location it also makes it more difficult (and less likely) for other applications to interoperate successfully with nushell (e.g. config file discovery).

**Reply by fdncred** [reply]
This is why you're able to change it.
https://github.com/nushell/nushell/discussions/5279#discussioncomment-4460912
and on discord something a little different
https://discord.com/channels/601130461678272522/1251882170658852914/1253702453270941858

oscarvarto mentions

I struggled to find documentation/answers to this. But finally got a working solution using macOS property list files. I ended up creating a me.oscarvarto.environment.plist file at the user level in my ~/Library/LaunchAgents/ folder.
... xml below is here ...
This is read by my system and helped me get my configuration in  ~/.config/nushell (a convenient dir name without whitespace for me, instead of the default that uses whitespace and breaks some nu stuff).

Leaving this for posterity, in case someone else is in the same need 🙂 

```
> cat me.oscarvarto.environment.plist
```
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>me.oscarvarto.environment</string>
    <key>ProgramArguments</key>
    <array>
        <string>/bin/sh</string>
        <string>-c</string>
        <string>
            launchctl setenv XDG_CONFIG_HOME /Users/oscarvarto/.config &&
            launchctl setenv XDG_CACHE_HOME /Users/oscarvarto/.cache &&
            launchctl setenv XDG_DATA_HOME /Users/oscarvarto/.local/share &&
            launchctl setenv XDG_STATE_HOME /Users/oscarvarto/.local/state &&
            launchctl setenv XDG_RUNTIME_DIR /Users/oscarvarto/.local/run &&
            launchctl setenv XDG_BIN_HOME /Users/oscarvarto/.local/bin
        </string>
    </array>
    <key>RunAtLoad</key>
    <true/>
    <key>ServiceIPC</key>
    <false/>
</dict>
</plist>
```

**Reply by tmillr** [reply]
Thanks for sharing, but you see this ***forces*** alot of users to have to go out of their way to setup these `XDG_*` variables (often to set them globally) ***for the first time*** in many cases (myself included), whereas almost every other terminal/unix-tool-like application simply "just works" how you expect it to (i.e. follows the XDG spec if/when these variables are unset, thereby utilizing the proper defaults).

Manually setting these variables up to their ***implicit defaults***, just because a single popular terminal/unix-like tool decides not to follow the spec, makes little sense (on unix-like systems where there is a well-known/well-defined default).

But anyway it's w/e, it's not a _huge_ deal I guess. It's just not ideal nor the norm when it comes to these kinds of tools.

**Reply by Mrfiregem** [reply]
Another reason a `.plist` file should be considered a hack besides being difficult for casual users to figure out, is that they aren't loaded before the window manager. If you have your terminal auto-open on startup, the variables will most likely not yet be set, so you'd have to quit and reopen it anyway.
